File: Solutions.java/Backtracking/0131. Palindrome Partitioning.java
/// <summary>
/// 131. Palindrome Partitioning
/// https://leetcode.com/problems/palindrome-partitioning
/// 
/// Difficulty Medium
/// Acceptance 69.9%
/// 
/// String
/// Dynamic Programming
/// Backtracking
/// </summary>
class Solution {
    public List<List<String>> partition(String s) {
        this.s = s;
        result = new ArrayList<List<String>>();
        len = s.length();
        pal = new int[len + 1][len + 1];
 
        Check(new ArrayList<String>(), 0);

        return result;
    }

    private String s;
    private List<List<String>> result;
    private int len;
    private int[][] pal;

    private void Check(ArrayList<String> row, int start) {
        if (IsPalindrome(start, len)) {
            var cloned = new ArrayList<String>(row);
            cloned.add(s.substring(start, len));
            result.add(cloned);
        }

        for (var i = start + 1; i < len; i++) {
            if (IsPalindrome(start, i)) {
                var cloned = new ArrayList<String>(row);
                cloned.add(s.substring(start, i));
                Check(cloned, i);
            }
        }
    }

    private Boolean IsPalindrome(int start, int end) {
        if (pal[start][end] != 0) {
            return pal[start][end] == 1;
        }

        var p = IsPalindromeInternal(start, end);
        pal[start][end] = p ? 1 : -1;
        return p;
    }

    private Boolean IsPalindromeInternal(int start, int end) {
        end--;
        while (start < end) {
            if (s.charAt(start) != s.charAt(end)) {
                return false;
            }
            start++;
            end--;
        }

        return true;
    }
}

File: Solutions.cs/Backtracking/0131. Palindrome Partitioning.cs
namespace Problem131
{

  /// <summary>
  /// 131. Palindrome Partitioning
  /// https://leetcode.com/problems/palindrome-partitioning
  /// 
  /// Difficulty Medium
  /// Acceptance 69.9%
  /// 
  /// String
  /// Dynamic Programming
  /// Backtracking
  /// </summary>
  public class Solution
  {
    public IList<IList<string>> Partition(string s)
    {
      S = s;
      Len = s.Length;
      Pal = new int[Len + 1, Len + 1];

      Check([], 0);

      return Result.Select(l => (IList<string>)l).ToList();
    }

    private string S;
    private List<List<string>> Result = new();
    private int Len;
    private int[,] Pal;

    private void Check(List<string> row, int start)
    {
      if (IsPalindrome(start, Len))
      {
        var cloned = new List<string>(row);
        cloned.Add(S.Substring(start, Len - start));
        Result.Add(cloned);
      }

      for (var i = start + 1; i < Len; i++)
      {
        if (IsPalindrome(start, i))
        {
          var cloned = new List<string>(row);
          cloned.Add(S.Substring(start, i - start));
          Check(cloned, i);
        }
      }
    }

    private bool IsPalindrome(int start, int end)
    {
      if (Pal[start, end] != 0)
        return Pal[start, end] == 1;

      var p = IsPalindromeInternal(start, end);
      Pal[start, end] = p ? 1 : -1;
      return p;
    }

    private bool IsPalindromeInternal(int start, int end)
    {
      end--;
      while (start < end)
      {
        if (S[start] != S[end])
          return false;
        start++;
        end--;
      }

      return true;
    }
  }
}
